File: projekt6/prio_czyt/main.py
#!/bin/python3

import multiprocessing
from os import getpid, popen, wait
import random
import sys
from time import sleep
space = 0
writerSem = multiprocessing.Semaphore(1)
readerSem = multiprocessing.Semaphore(1)
mem = multiprocessing.Value('c')
lc = multiprocessing.Value('i', 0)
sp = multiprocessing.Value('i', 0)
def writer():
    while(1):
        writerSem.acquire()
        if sp.value - space == 0:
            print("Writer "+str(getpid())+" come to library")
            mem.value = random.randint(97,122)
            # print("Writer "+str(getpid())+" out of library")
            sleep(0.1)
        writerSem.release()

def reader():
    while(1):
        readerSem.acquire()
        if sp.value - 1 >= 0:
            sp.value -= 1
            lc.value += 1
            if lc.value == 1:
                writerSem.acquire()
            # print("Reader "+str(getpid())+" come to library")
            readerSem.release()
            x = mem.value
            print("Reader "+str(getpid())+" "+x.decode("utf-8")+ " "+str(sp.value))
            readerSem.acquire()
            # print("Reader "+str(getpid())+" out of library")
            lc.value -= 1
            if lc.value == 0:
                writerSem.release()
            sp.value += 1
            sleep(0.1)
        readerSem.release()




name = sys.argv[0]
# print(str(getpid()))
if len(sys.argv) != 4:
    print("Nieodpowiednia liczba argumentów")
    exit(-1)

if int(sys.argv[1]) <= 0 or int(sys.argv[2]) <= 0 or int(sys.argv[3]) < 0:
    print("Wartosci argumentow sa nieodpowiednie")
    exit(-1)

numOfWriters = int(sys.argv[1])
numOfReaders = int(sys.argv[2])
space = int(sys.argv[3])
try:
    process = popen('ps ux | wc -l', 'r')
    processLimit = popen('ulimit -p', 'r')
except Exception as e:
    print(name+": ", end = "")
    print(e)
    exit(-1)

if int(process.readline())-4 + numOfWriters + numOfReaders> int(processLimit.readline()):
    print("Przekroczono limit procesow")
    exit(-1)
try:
    process.close()
    processLimit.close()
except Exception as e:
    print(name+": ", end = "")
    print(e)
    exit(-1)

# numOfReaders = 1
# numOfWriters = 1
sp.value = space
for i in range(0, numOfReaders):
    p1 = multiprocessing.Process(target=reader)
    p1.start()

for i in range(0, numOfWriters):
    p2 = multiprocessing.Process(target=writer)
    p2.start()

wait()

File: projekt6/prio_czyt/main.cpp
#include <sys/wait.h>
#include <signal.h>
#include <iostream>
#include <sys/errno.h>
#include <sys/types.h> 
#include <sys/ipc.h>
#include <sys/shm.h>
#include <sys/sem.h>
#include <unistd.h>
#include <vector>
using namespace std;
vector<int> ppid;
int semId, shmId;
long numOfWriters, numOfReaders, space;
void del(){
    try {
        semctl(semId, 0, IPC_RMID);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    try {
        shmctl(shmId, IPC_RMID, 0);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
}
void interruption(int signum)
{
    for(auto p : ppid){
        kill(p, SIGINT);
    }
    del();
    exit(signum);
}
int main(int argc, char* argv[]) {
    int i, *adress;
    char *end;
    
    void interruption(int signum);
    signal(SIGINT, interruption);
    if(argc != 4){
        cout << "Main: Nieodpowiednia liczba argumentow" << endl;
        exit(-1);
    }
    errno = 0;
    try {
        numOfWriters = strtol(argv[1], NULL, 10);
        numOfReaders = strtol(argv[2], NULL,10);
        space = strtol(argv[3], NULL, 10);
        if (errno != 0){
            throw runtime_error("Main: Blad strtol");
        }
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    if(numOfWriters <=0  || numOfReaders <= 0 || space <= 0) {
        cout << "Main: Nieprawidlowe wartosci podane w argumentach Main, powinny byc 3 argumenty o liczbach dodatnich!" << endl;
        exit(-1);
    }
    
    FILE *pd, *pd1;
    int processLimit, numOfProcess;
    try {
        pd = popen("ps ux | wc -l", "r");
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    try {
        fscanf(pd, "%d", &numOfProcess);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    try {
        pclose(pd);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }

    try {
        pd = popen("ulimit -p", "r");
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    try {
        fscanf(pd, "%d", &processLimit);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }

    numOfProcess -= 4;
    try {
        pclose(pd);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }

    if (numOfProcess + numOfReaders + numOfWriters > processLimit) {
        cout << "Przekroczono limit procesów" << endl;
        exit(-1);
    }

    try {
        semId = semget(997, 2, IPC_CREAT|0606);
    } catch (exception &e) {
        cerr << e.what() << endl;
        exit(-1);
    }
    for (i = 0; i < 2; i++){
        try {
            semctl(semId, i, SETVAL, 1);
        } catch (exception &e) {
            cerr << e.what() << endl;
            exit(-1);
        }
    }
    try {
        shmId = shmget(997, sizeof(int), IPC_CREAT|0606);
    } catch (exception &e) {
            cerr << e.what() << endl;
            exit(-1);
    }
    try {
        adress = (int*)shmat(shmId, 0, 0);
        adress[0] = 0;
        shmdt(adress);
    } catch (exception &e) {
            cerr << e.what() << endl;
            exit(-1);
    }

    int x;
    for(i = 0; i < numOfWriters; i++){
        switch (x = fork())
        {
        case -1:
            cerr << "Main: Blad podczas tworzenia fork() dla piszarza";    
        case 0:
            try {
                execl("pisarz","/pisarz", argv[3], NULL);
            } catch (exception &e) {
                del();
                cerr << "Main: Blad podczas tworzenia exec() dla piszarza";
                exit(-1);
            }
        default:
            ppid.push_back(x);
        }
    }

    for(i = 0; i < numOfReaders; i++) {
        switch (x = fork())
        {
        case -1:
            cerr << "Main: Blad podczas tworzenia fork() dla czytelnika";    
        case 0:
            try {
                execl("czytelnik","./czytelnik", argv[3], NULL);
            } catch (exception &e) {
                cerr << "Main: Blad podczas tworzenia exec() dla czytelnika";
                exit(-1);
            }
        default:
            ppid.push_back(x);
        }
    }

    wait(NULL);
    interruption(SIGINT);
}